### PR TITLE
[COS-834] - COS API Create Contract 

### PR DESCRIPTION
**Pull request by grokkos**
Create Contract COS API layer implementation.

- [ ] TODO the test on `contract.resolvers_it_test.go`

## Changes

What types of changes does your code introduce?  _Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build related changes
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Other (please describe below)

## Summary by CodeRabbit

- **New Features**
  - Introduced a new contract management system, including the ability to create contracts and view their status and renewal cycles.
  - Added a search bar to the main interface for quick access to search functionality.
  - Implemented a new GraphQL schema for contracts, allowing for advanced querying and mutation operations.

- **Enhancements**
  - Improved user interface with new styles for the search bar, enhancing the overall user experience.

- **Bug Fixes**
  - Addressed issues where certain contract fields were not resolving correctly by implementing new resolver functions.

- **Documentation**
  - Updated GraphQL schema documentation to reflect new contract-related types, inputs, and enums.

- **Tests**
  - Added new integration tests for the contract creation mutation to ensure reliability and performance.

- **Chores**
  - Performed general codebase maintenance to keep the code clean and well-organized.

- **Refactor**
  - Refactored contract-related code to improve maintainability and scalability as new features are added.

### CODE REVIEW

**Comment by grokkos** [comment]
@alexopenline I need to double check with you, the contract's `entity` fields and the `neo4jDb` and `cypher` queries. 
If you observe something out of context, please let me know🙏

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    renewalCycle: ContractRenewalCycle!
+    status: ContractStatus!
+    owner: User @goField(forceResolver: true)
+    source: DataSource!
+    appSource: String!
+    externalLinks: [ExternalSystem!]! @goField(forceResolver: true)
+    contractUrl:        String!
+}
+
+input ContractInput {
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls:27]: pls add contractUrl string, optional

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    referenceId:   String
+    name:          String!
+    organizationId: String!
+    renewalCycle: ContractRenewalCycle!
+    status: ContractStatus!
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls]: in input can be optional, in create API will be derived

File: packages/server/customer-os-api/entity/contract.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package entity
+
+import (
+	"time"
+)
+
+type ContractEntity struct {
```
Comment (alexopenline) [inline packages/server/customer-os-api/entity/contract.go:7]: add contractUrl

File: packages/server/customer-os-api/entity/contract.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	ID                                 string
+	CustomerOsId                       string     `neo4jDb:"property:customerOsId;lookupName:CUSTOMER_OS_ID;supportCaseSensitive:false"`
+	ReferenceId                        string     `neo4jDb:"property:referenceId;lookupName:REFERENCE_ID;supportCaseSensitive:true"`
+	Name                               string     `neo4jDb:"property:name;lookupName:NAME;supportCaseSensitive:true"`
+	CreatedAt                          time.Time  `neo4jDb:"property:createdAt;lookupName:CREATED_AT;supportCaseSensitive:false"`
+	UpdatedAt                          time.Time  `neo4jDb:"property:updatedAt;lookupName:UPDATED_AT;supportCaseSensitive:false"`
+	ServiceStartedAt                   *time.Time `neo4jDb:"property:serviceStartedAt;lookupName:SERVICE_STARTED_AT;supportCaseSensitive:false"`
+	ServiceStartedId                   *string    `neo4jDb:"property:serviceStartedId;lookupName:SERVICE_STARTED_ID;supportCaseSensitive:false"`
```
Comment (alexopenline) [inline packages/server/customer-os-api/entity/contract.go]: Do not have this field in entity in DB

File: packages/server/customer-os-api/entity/contract.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package entity
+
+import (
+	"time"
+)
+
+type ContractEntity struct {
+	ID                                 string
+	CustomerOsId                       string     `neo4jDb:"property:customerOsId;lookupName:CUSTOMER_OS_ID;supportCaseSensitive:false"`
```
Comment (alexopenline) [inline packages/server/customer-os-api/entity/contract.go]: Not needed yet.

File: packages/server/customer-os-api/entity/contract.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	ReferenceId                        string     `neo4jDb:"property:referenceId;lookupName:REFERENCE_ID;supportCaseSensitive:true"`
```
Comment (alexopenline) [inline packages/server/customer-os-api/entity/contract.go]: Not needed yet

File: packages/server/customer-os-api/entity/contract.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	ServiceStartedAt                   *time.Time `neo4jDb:"property:serviceStartedAt;lookupName:SERVICE_STARTED_AT;supportCaseSensitive:false"`
+	ServiceStartedId                   *string    `neo4jDb:"property:serviceStartedId;lookupName:SERVICE_STARTED_ID;supportCaseSensitive:false"`
+	ServiceSignedAt                    *time.Time `neo4jDb:"property:serviceSignedAt;lookupName:SERVICE_SIGNED_AT;supportCaseSensitive:false"`
+	ServiceSignedId                    *string    `neo4jDb:"property:serviceSignedId;lookupName:SERVICE_SIGNED_ID;supportCaseSensitive:false"`
+	ContractRenewalCycle               ContractRenewalCycle
+	ContractStatus                     ContractStatus
+	Source                             DataSource
+	AppSource                          string
+	ContractUrl                        string
+	InteractionEventParticipantDetails InteractionEventParticipantDetails
```
Comment (alexopenline) [inline packages/server/customer-os-api/entity/contract.go]: Not needed

File: packages/server/customer-os-api/entity/contract.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+	ReferenceId                        string     `neo4jDb:"property:referenceId;lookupName:REFERENCE_ID;supportCaseSensitive:true"`
+	Name                               string     `neo4jDb:"property:name;lookupName:NAME;supportCaseSensitive:true"`
+	CreatedAt                          time.Time  `neo4jDb:"property:createdAt;lookupName:CREATED_AT;supportCaseSensitive:false"`
+	UpdatedAt                          time.Time  `neo4jDb:"property:updatedAt;lookupName:UPDATED_AT;supportCaseSensitive:false"`
+	ServiceStartedAt                   *time.Time `neo4jDb:"property:serviceStartedAt;lookupName:SERVICE_STARTED_AT;supportCaseSensitive:false"`
+	ServiceStartedId                   *string    `neo4jDb:"property:serviceStartedId;lookupName:SERVICE_STARTED_ID;supportCaseSensitive:false"`
+	ServiceSignedAt                    *time.Time `neo4jDb:"property:serviceSignedAt;lookupName:SERVICE_SIGNED_AT;supportCaseSensitive:false"`
+	ServiceSignedId                    *string    `neo4jDb:"property:serviceSignedId;lookupName:SERVICE_SIGNED_ID;supportCaseSensitive:false"`
```
Comment (alexopenline) [inline packages/server/customer-os-api/entity/contract.go]: do not have this field in db

File: packages/server/customer-os-api/entity/contract.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+type ContractEntity struct {
+	ID                                 string
+	CustomerOsId                       string     `neo4jDb:"property:customerOsId;lookupName:CUSTOMER_OS_ID;supportCaseSensitive:false"`
+	ReferenceId                        string     `neo4jDb:"property:referenceId;lookupName:REFERENCE_ID;supportCaseSensitive:true"`
+	Name                               string     `neo4jDb:"property:name;lookupName:NAME;supportCaseSensitive:true"`
+	CreatedAt                          time.Time  `neo4jDb:"property:createdAt;lookupName:CREATED_AT;supportCaseSensitive:false"`
+	UpdatedAt                          time.Time  `neo4jDb:"property:updatedAt;lookupName:UPDATED_AT;supportCaseSensitive:false"`
+	ServiceStartedAt                   *time.Time `neo4jDb:"property:serviceStartedAt;lookupName:SERVICE_STARTED_AT;supportCaseSensitive:false"`
+	ServiceStartedId                   *string    `neo4jDb:"property:serviceStartedId;lookupName:SERVICE_STARTED_ID;supportCaseSensitive:false"`
+	ServiceSignedAt                    *time.Time `neo4jDb:"property:serviceSignedAt;lookupName:SERVICE_SIGNED_AT;supportCaseSensitive:false"`
```
Comment (alexopenline) [inline packages/server/customer-os-api/entity/contract.go]: Let's renamed to SignedAt, as it refers to whole contract,
Also pls add EndedAt *time field

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 10]
```diff
@@ -0,0 +1,54 @@
+extend type Query {
+    contract(id: ID!): Contract! @hasRole(roles: [ADMIN, USER]) @hasTenant
+}
+extend type Mutation {
+    contract_Create(input: ContractInput!): Contract!  @hasRole(roles: [ADMIN, OWNER]) @hasTenant
+}
+
+type Contract implements Node {
+    id: ID!
+    referenceId:   String
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls]: remove

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    customerOsId: String!
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls]: remove

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    organizationId: String!
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls]: should be a reference to contract object, or let's exclude for now

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    createdAt:   Time!
+    updatedAt:   Time!
+
+    serviceStartedAt:   Time!
+    serviceStartedAtTimelineEventId: ID #we need this in order to use the dataloader for the serviceStartedAtTimelineEvent if asked
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls]: timeline events will be done differently these 2 fields can be removed

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    serviceStartedAtTimelineEvent: TimelineEvent @goField(forceResolver: true)
+
+    serviceSignedAt:   Time!
+    serviceSignedAtTimelineEventId: ID #we need this in order to use the dataloader for the serviceSignedAtTimelineEvent if asked
+    serviceSignedAtTimelineEvent: TimelineEvent @goField(forceResolver: true)
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls]: same as started at

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+    name:        String!
+    sourceFields: SourceFields!
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls]: there is a duplication between sourceFields and Datasource

File: packages/server/customer-os-api/graph/schemas/contract.graphqls [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    appSource:     String
+}
+
+enum ContractRenewalCycle {
+    NONE
+    MONTHLY_RENEWAL
+    ANNUAL_RENEWAL
+}
+enum ContractStatus {
+    UNDEFINED
```
Comment (alexopenline) [inline packages/server/customer-os-api/graph/schemas/contract.graphqls:50]: Can be removed

File: packages/server/customer-os-api/repository/contract_repository.go [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+	driver *neo4j.DriverWithContext
+}
+
+func NewContractRepository(driver *neo4j.DriverWithContext) ContractRepository {
+	return &contractRepository{
+		driver: driver,
+	}
+}
+
+func (r *contractRepository) CreateContract(ctx context.Context, tenant, organizationId string, entity entity.ContractEntity) (*dbtype.Node, error) {
```
Comment (alexopenline) [inline packages/server/customer-os-api/repository/contract_repository.go]: Contract creation is done via events, 
need to call corresponding event via grpc.
This repository method to be removed

File: packages/server/customer-os-api/repository/contract_repository.go [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+			})
+		return utils.ExtractSingleRecordFirstValueAsNode(ctx, queryResult, err)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return result.(*dbtype.Node), nil
+}
+
+func (r *contractRepository) SetContractCreator(ctx context.Context, tenant, userId, contractId string) error {
```
Comment (alexopenline) [inline packages/server/customer-os-api/repository/contract_repository.go]: to be removed, creator is set in events

File: packages/server/customer-os-api/service/contract_service.go [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+type ContractCreateData struct {
+	ContractEntity    *entity.ContractEntity
+	Organization      string
+	ExternalReference *entity.ExternalSystemEntity
+	Source            entity.DataSource
+	AppSource         string
+}
+
+func (s *contractService) CreateContract(ctx context.Context, organization string, entity *entity.ContractEntity) (*entity.ContractEntity, error) {
```
Comment (alexopenline) [inline packages/server/customer-os-api/service/contract_service.go]: in this method, just invoke event from events-processing-platform.

**Review by alexopenline — CHANGES_REQUESTED** [review]

**Review by alexopenline — APPROVED** [review]

**Pull request merged**